src/api/uploads/MultiputPart.js

```diff
@@ -34,30 +51,264 @@ class MultiputPart extends Base {
+            return;
+        }
+
+        const clientEventInfo = {
+            documentHidden: document.hidden,
+            digest_retries: this.numDigestRetriesPerformed,
+            timing: this.timing,
+            parts_uploading: this.getNumPartsUploading()
+        };
+
+        const headers = {
+            'Content-Type': 'application/octet-stream',
+            Digest: `sha-256=${this.sha256}`,
+            'Content-Range': `bytes ${this.offset}-${this.rangeEnd}/${this.size}`,
+            'X-Box-Client-Event-Info': JSON.stringify(clientEventInfo)
+        };
+
+        this.startTimestamp = Date.now();
+
+        this.xhr.uploadFile({
+            url: this.uploadUrl,
+            data: this.blob,
+            headers,
+            method: 'PUT',
+            successHandler: this.uploadSuccessHandler,
+            errorHandler: this.uploadErrorHandler,
+            progressHandler: this.uploadProgressHandler,
+            withIdleTimeout: true,
+            idleTimeoutDuration: this.config.requestTimeoutMS
+        });
+    };
+
+    /**
+     * Handler for upload part success
+     * 
+     * @param {Object} data
+     * @return {void}
+     */
+    uploadSuccessHandler = (data: Object) => {
+        if (this.isDestroyed()) {
+            return;
+        }
+        this.state = PART_STATE_UPLOADED;
+        this.consoleLogFunc(() => `Upload completed: ${this.stringify()}. Parts state: ${this.getPartsState()}`);
+        this.data = data;
+        this.blob = null;
+        this.timing.uploadTime = Date.now() - this.startTimestamp;
+
+        this.onSuccess(this);
+
+        this.uploadedBytes = this.size;
+    };
+
+    /**
+     * Handler for upload part progress event
+     * 
+     * @param {ProgressEvent} data
+     * @return {void}
+     */
+    uploadProgressHandler = (event: ProgressEvent) => {
+        const newUploadedBytes = parseInt(event.loaded, 10);
+        this.onProgress(this.uploadedBytes, newUploadedBytes);
+        this.uploadedBytes = newUploadedBytes;
+    };
+
     /**
-     * Upload the part
-     * TODO: implement this
+     * Handler for upload part error
      * 
-     * @public
+     * @param {Error} error
      * @return {void}
      */
-    upload = (): void => {};
+    uploadErrorHandler = (error: Error) => {
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        this.consoleLogFunc(
+            () =>
+                `Upload failure ${error.message} for part ${this.stringify()}. XHR state: ${this.xhr.xhr
+                    .readyState}. Parts state ${this.getPartsState()}`
+        );
+        const eventInfo = {
+            message: error.message,
+            part: {
+                uploadedBytes: this.uploadedBytes,
+                id: this.id,
+                index: this.index,
+                offset: this.offset
+            },
+            xhr_ready_state: this.xhr.xhr.readyState,
+            xhr_status_text: this.xhr.xhr.statusText
```

Comment:
Is this just the event format directly ported over from WebApp code?

Reply:
yeah